prov/gni: Implemented fi direct.

e-harvey opened the pull request:
- Added missing gnix functions.
- Added DIRECT_FN and STATIC to the beginning of GNI
API functions to make those functions globally visible
when compiling with fi direct enabled.
- Updated Makefile.include to include gnix_poll.*.
- Added symbol version info for the gni provider.

Signed-off-by: Evan Harvey <eharvey@lanl.gov>

@hppritcha @sungeunchoi @jswaro @jshimek @tonyzinger @chuckfossen 

Fixes #600

lanl-ompi commented:
Test FAILed.

e-harvey commented:
Perhaps we could add '**attribute**(always_inline)' to the 'DIRECT_FN' macro defined in fi_prov.h.  Will this force inlining when compiling with clang?

hppritcha commented:
not sure, but probably its worth adding this attribute to the function declarations.

sungeunchoi commented:
Yes newer version support always_inline (don't know how new it has to be).  That said, we're not building with clang on our machines and currently don't have any plans to do so.

hppritcha commented:
@e-harvey i renamed the include file from the provider that goes in to libfabric.map in response to sean's concerns about ofiwg/libfabric#1740

e-harvey commented:
@hppritcha Ok, I renamed provider.map -> provider_FABRIC_1.0.map.

On Wed, 24 Feb 2016 10:04:59 -0800
Howard Pritchard notifications@github.com wrote:

> @e-harvey i renamed the include file from the provider that goes in to libfabric.map in response to sean's concerns about ofiwg/libfabric#1740
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ofi-cray/libfabric-cray/pull/671#issuecomment-188385232

## 

Evan Harvey

e-harvey commented:
Rebased with maser, moved DIRECT_FN and STATIC macros in gnix.h, and added gnix_fabric_trywait.  Please review.

hppritcha commented on prov/gni/src/gnix_ep.c:
```diff
@@ -634,20 +655,43 @@ static ssize_t gnix_ep_tsendmsg(struct fid_ep *ep,
+ * @return FI_SUCCESS	upon successfully writing to the destination
+ * @return -FI_ERRNO	upon an error
+ * @return -FI_ENOSYS	if this operation is not supported
+ */
```
thanks for adding the doxygen style comments

hppritcha commented:
:+1:
